File: exact_diag.jl
using LinearAlgebra
using Statistics
using DelimitedFiles

# Problem setup: 1×8 black/white image
nrows, ncols = 1, 8
N = nrows*ncols
bw = vcat(zeros(Float64, 4), ones(Float64, 4))  # 0..1
img = reshape(bw, nrows, ncols)
pix = vec(img) .* 255.0
σ = std(pix; corrected=false)

# Couplings J_x per horizontal bond: J = 2 - (4/σ)*|ΔI|
Jx = zeros(N-1)
for c in 1:(N-1)
    Δ = abs(pix[c] - pix[c+1])
    Jx[c] = 2.0 - (4.0/σ)*Δ
end

# Optional seeds: list of (site, label, strength) for local Sx fields
seeds = Tuple{Int,Float64,Float64}[]  # e.g., push!(seeds, (2, -1.0, 2.0))

# Single-site spin-1/2 operators (Sz basis)
σx = [0 1; 1 0];  σy = [0 -im; im 0];  σz = [1 0; 0 -1]
Sx = 0.5*σx;      Sy = 0.5*σy;          Sz = 0.5*σz
I2 = Matrix{ComplexF64}(I, 2, 2)

# Place local and two-site operators into the full 2^N space
function op_on(i::Int, A::AbstractMatrix{<:Complex}, N::Int)
    T = (i==1 ? A : I2)
    for s in 2:N
        T = kron(T, s==i ? A : I2)
    end
    return T
end

function op2_on(i::Int, j::Int, A::AbstractMatrix{<:Complex}, B::AbstractMatrix{<:Complex}, N::Int)
    @assert 1 <= i < j <= N
    T = (i==1 ? A : I2)
    for s in 2:N
        if s == j
            T = kron(T, B)
        elseif s == i
            # already placed A
        else
            T = kron(T, I2)
        end
    end
    return T
end

# Build dense Hamiltonian: H = sum_bonds (-SySy - SzSz - Jx*SxSx) + sum_seeds (-h*label*Sx)
dim = 2^N
H = zeros(ComplexF64, dim, dim)

for c in 1:(N-1)
    H .+= -op2_on(c, c+1, Sy, Sy, N)
    H .+= -op2_on(c, c+1, Sz, Sz, N)
    H .+= -Jx[c] * op2_on(c, c+1, Sx, Sx, N)
end

for (site, label, strength) in seeds
    H .+= -(strength*label) * op_on(site, Sx, N)
end

# Exact diagonalization
F = eigen(Hermitian(H))
evals = F.values
V = F.vectors  # columns are eigenvectors |E_α>

# Initial states in Sz basis (dense vectors)

# Uniform over all configurations: |+x>^⊗N
function plusx_vec(N::Int)
    inv√2 = inv(sqrt(2.0))
    v1 = ComplexF64[inv√2, inv√2]
    v = v1
    for _ in 2:N
        v = kron(v, v1)
    end
    v ./= norm(v)
    return v
end

# Equal superposition of all basis states with exactly M up-spins
function fixed_upspin_vec(N::Int, M::Int)
    v = zeros(ComplexF64, 2^N)
    for b in 0:(UInt(1)<<N)-1
        if count_ones(b) == M
            idx = Int(b) + 1  # Sz basis with site 1 as least-significant bit, |↑> mapped to bit=1
            v[idx] = 1
        end
    end
    if norm(v) > 0
        v ./= norm(v)
    end
    return v
end

v_plusx = plusx_vec(N)
M_up = round(Int, N * mean(img))
v_fixed = fixed_upspin_vec(N, M_up)

# Expansion coefficients |c_α|^2 for both initial states
coeffs_plusx = abs2.(conj(V)' * v_plusx)
coeffs_fixed = abs2.(conj(V)' * v_fixed)

println("N = $N, mean(image) = $(mean(img)), M_up = $M_up")
println("sum |c_α|^2 (plusx) = ", sum(coeffs_plusx))
println("sum |c_α|^2 (fixedM) = ", sum(coeffs_fixed))

# Save results
mkpath("examples")
writedlm("examples/ed_evals.csv", evals, ',')
writedlm("examples/ed_coeffs_plusx.csv", coeffs_plusx, ',')
writedlm("examples/ed_coeffs_fixedM$(M_up).csv", coeffs_fixed, ',')

# Print top contributors
perm_plusx = sortperm(coeffs_plusx; rev=true)
perm_fixed = sortperm(coeffs_fixed; rev=true)
println("top 5 (plusx): ", [(perm_plusx[k], coeffs_plusx[perm_plusx[k]]) for k in 1:min(5, length(perm_plusx))])
println("top 5 (fixedM): ", [(perm_fixed[k], coeffs_fixed[perm_fixed[k]]) for k in 1:min(5, length(perm_fixed))])

File: exact_diag.py
import numpy as np
from numpy.linalg import eigh
from functools import reduce
import matplotlib.pyplot as plt

# Problem setup: 1×8 black/white "image"
n = 8
img01 = np.concatenate([np.zeros(4), np.ones(4)])  # first 4 black, last 4 white
pix = img01 * 255.0
sigma = pix.std(ddof=0)

# Couplings J_x
Delta = np.zeros(n - 1)
for i in range(n - 1):
    d = abs(pix[i] - pix[i + 1])
    Delta[i] = 2.0 - (4.0 / sigma) * d
print("Delta=",Delta)

# Seeds (empty by default)
seeds = [(0,1,10),(7,-1,10)]

# Single-site operators
sx = np.array([[0, 1], [1, 0]], dtype=np.complex128)
sy = np.array([[0, 1j], [-1j, 0]], dtype=np.complex128)
sz = np.array([[1, 0], [0, -1]], dtype=np.complex128)
Sx = 0.5 * sx
Sy = 0.5 * sy
Sz = 0.5 * sz
I2 = np.eye(2, dtype=np.complex128)

# Helpers to embed operators
def kron_all(mats):
    return reduce(np.kron, mats)

def op_on(site, A):
    mats = [I2] * n
    mats[site] = A
    return kron_all(mats)

def op2_on(i, j, A, B):
    mats = [I2] * n
    mats[i] = A
    mats[j] = B
    return kron_all(mats)

# Build Hamiltonian
dim = 2 ** n
H = np.zeros((dim, dim), dtype=np.complex128)

for i in range(n - 1):
    H -= op2_on(i, i + 1, Sy, Sy)
    H -= op2_on(i, i + 1, Sz, Sz)
    H -= Jx[i] * op2_on(i, i + 1, Sx, Sx)

for (site, label, strength) in seeds:
    H -= (strength * label) * op_on(site, Sx)



# Diagonalize
evals, V = eigh(H)

# Initial states
def plusx_vec(n):
    v1 = np.array([1/np.sqrt(2), 1/np.sqrt(2)], dtype=np.complex128)
    v = v1
    for _ in range(1, n):
        v = np.kron(v, v1)
    v /= np.linalg.norm(v)
    return v

def fixed_upspin_vec(n, M):
    v = np.zeros(2**n, dtype=np.complex128)
    for b in range(2**n):
        if bin(b).count("1") == M:
            v[b] = 1.0
    if np.linalg.norm(v) > 0:
        v /= np.linalg.norm(v)
    return v

v_plusx = plusx_vec(n)
M_up = int(round(n * img01.mean()))
v_fixed = fixed_upspin_vec(n, M_up)

# Expansion coefficients
c2_plusx = np.abs(V.conj().T @ v_plusx) ** 2
c2_fixed = np.abs(V.conj().T @ v_fixed) ** 2

print(f"N = {n}, mean(image) = {img01.mean():.3f}, M_up = {M_up}")
print("sum |c_alpha|^2 (plusx) =", float(c2_plusx.sum()))
print("sum |c_alpha|^2 (fixedM) =", float(c2_fixed.sum()))

# Top contributors
topk = 20
idx_plusx = np.argsort(-c2_plusx)[:topk]
idx_fixed = np.argsort(-c2_fixed)[:topk]
#print("top 5 (plusx):", [(int(i), float(c2_plusx[i])) for i in idx_plusx])
print("top 5 (fixedM):", [(int(i), float(c2_fixed[i])) for i in idx_fixed])
print("Probability mass (top 10):", sum([float(c2_fixed[i]) for i in idx_fixed]))

# print("V(70) = ", V[70])
print("Ground state:", np.argmin(evals))

topkstates = [V[:,idx_fixed[i]] for i in range(topk)]
#state1 = V[:,idx_fixed[0]]
#state2 = V[:,idx_fixed[1]]
#state3 = V[:,idx_fixed[2]]
#state4 = V[:,idx_fixed[3]]
#state5 = V[:,idx_fixed[4]]
ground = V[:,0]

expect_sx = np.zeros(n)
expect = [np.zeros(n) for _ in range(topk)]

for j in range(n):
    expect_sx[j] = np.real(np.vdot(ground,op_on(j,Sx)@ground))
    for i in range(topk):
        expect[i][j] = np.real(np.vdot(topkstates[i],op2_on(0,j,Sx,Sx)@topkstates[i]))

    

print("Expected spin (ground):", expect_sx)
#print("Expected spin (state1):", expect_sx1)
#print("Expected spin (state2):", expect_sx2)
#print("Expected spin (state3):", expect_sx3)
#print("Expected spin (state4):", expect_sx4)
#print("Expected spin (state5):", expect_sx5)

weighted_spin = sum([float(c2_fixed[idx_fixed[i]])*expect[i] for i in range(topk)])

print("Weighted expected spin (top 5):", weighted_spin)


# Plot overlap distributions
plt.figure(figsize=(10, 5))

plt.bar(range(len(c2_plusx)), c2_plusx, 
        alpha=0.6, label="plusx initial state")
plt.bar(range(len(c2_fixed)), c2_fixed, 
        alpha=0.6, label="fixed-M initial state")

# Mark the ground state index
ground_idx = np.argmin(evals)
plt.axvline(ground_idx, color="red", linestyle="--", label="Ground state")

plt.title("|c_alpha|^2 distribution over eigenstates")
plt.xlabel("Eigenstate index (α)")
plt.ylabel("|c_alpha|^2")
plt.legend()
plt.tight_layout()
plt.show()

# explore concentration of probability mass

# measure <S^x_j>, <S^x_0 S^x_j>, weighted sum of c_alpha